[Rando] Fix Check Tracker Area Totals

Malkierian opened the pull request:
After changing to the flags tracking system for the check tracker, it would show incorrect totals for checks gotten because invisible checks were getting set and adding to the count for that. Also, when adding "Always show gold skulltulas" option, the total checks shown in an area was wrong because all checks were being added to `checksByArea` all the time instead of only those visible. This fixes both, and both update in real time when "Always show gold skulltulas" or "Hide right-side shop item checks" are toggled.

Also reformats some of the code of `randomizer_check_tracker.cpp` to eliminate all remaining instances of unencapsulated if statements.

https://github.com/HarbourMasters/Shipwright/assets/774771/995dccaf-862f-4b03-8d7e-8c5a7412d953



<!--- section:artifacts:start -->
### Build Artifacts
  - [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1140091678.zip)
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1140091680.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1140091681.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1140091682.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1140091683.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1140091684.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1140091686.zip)
<!--- section:artifacts:end -->

Malkierian committed:
Fix the calculation of `areaChecksGotten` to account for flags functionality setting invisible checks.
Change `areaChecksTotal` to dynamic calculation based on tracker visibility, now that all checks are being added to `checksByArea`.
Both are updated in realtime when either "Show all GS locations" or "Hide right side shop items" are toggled.
Reformat all remaining unencapsulated if statements.

Malkierian commented:
https://github.com/HarbourMasters/Shipwright/assets/774771/8b6a62a3-430b-466c-9e93-d3562f5b42e6
WTF GitHub, why won't you embed this video?

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -977,11 +1014,17 @@ void CheckTrackerWindow::DrawElement() {
             for (auto rco : objs) {
-                if (doDraw && isThisAreaSpoiled && IsVisibleInCheckTracker(rco))
-                    DrawLocation(rco);
+                if (IsVisibleInCheckTracker(rco)) {
+                    if (doDraw && isThisAreaSpoiled) {
+                        DrawLocation(rco);
+                    }
+                } else {
+                    
+                }
```
is there a reason to have this split up like this with the empty else instead of just having

```suggestion
                if (IsVisibleInCheckTracker(rco) && doDraw && isThisAreaSpoiled)
                    DrawLocation(rco);
                }
```

Malkierian replied:
Oh, lol, I think that was leftover from an attempt to get it working a different way. I'll check to make sure I didn't miss implementing something and remove it if not.

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -787,9 +822,9 @@ void Teardown() {
+    if ((!gSaveContext.checkTrackerData[check].skipped && data.skipped) || (gSaveContext.checkTrackerData[check].status >= RCSHOW_COLLECTED && data.status < RCSHOW_COLLECTED)) {
         areaChecksGotten[area]++;
-    } else if (gSaveContext.checkTrackerData[check].skipped && !data.skipped) {
+    } else if ((gSaveContext.checkTrackerData[check].skipped && !data.skipped) || (gSaveContext.checkTrackerData[check].status < RCSHOW_COLLECTED && data.status >= RCSHOW_COLLECTED)) {
```
it's not really clear what `< RCSHOW_COLLECTED` and `>= RCSHOW_COLLECTED` means here. i think even if it's more verbose it would make sense to have this be

```suggestion
    } else if ((gSaveContext.checkTrackerData[check].skipped && !data.skipped) || ((gSaveContext.checkTrackerData[check].status != RCSHOW_COLLECTED && gSaveContext.checkTrackerData[check].status != RCSHOW_SAVED) && (data.status == RCSHOW_COLLECTED || data.status == RCSHOW_SAVED))) {
```

Malkierian replied:
Using >= or < here was just an attempt to shorten the code necessary. You're right that >= would mean it was either collected or saved and < is anything else. I don't particularly care either way, I just like more concise code that does the same job.

briaguya0 replied:
when reading `>=` i had no idea what it meant and had to go look up the enum to understand the line. so while it may be fewer characters it took much longer to understand

Malkierian replied:
To be fair, though, it wasn't an || for the comparison for == COLLECTED or SAVED, but an &&. And, I don't use this myself, Anchor was what kept this in there, and I don't have any way of testing it to make sure it's actually working like I expect XD.

briaguya0 replied:
> it wasn't an || for the comparison for == COLLECTED or SAVED, but an &&

i'm not sure what you mean here

```cpp
gSaveContext.checkTrackerData[check].status < RCSHOW_COLLECTED
```
is logically equivalent to
```cpp
gSaveContext.checkTrackerData[check].status != RCSHOW_COLLECTED && gSaveContext.checkTrackerData[check].status != RCSHOW_SAVED
```
and
```cpp
data.status >= RCSHOW_COLLECTED
```
is logically equivalent to
```cpp
data.status == RCSHOW_COLLECTED || data.status == RCSHOW_SAVED
```

Malkierian replied:
Oh, I misread your suggestion. I thought you were using || in the first instance for some reason.

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -515,8 +542,16 @@ void CheckTrackerTransition(uint32_t sceneNum) {
     if (IS_RANDO) {
-        hideShopRightChecks = CVarGetInteger("gCheckTrackerOptionHideRightShopChecks", 1);
-        alwaysShowGS = CVarGetInteger("gCheckTrackerOptionAlwaysShowGSLocs", 0);
+        bool hideShopRightChecks_ = CVarGetInteger("gCheckTrackerOptionHideRightShopChecks", 1);
+        if (hideShopRightChecks != hideShopRightChecks_) {
+            hideShopRightChecks = hideShopRightChecks_;
+            RecalculateAreaTotals();
+        }
```
would changing https://github.com/HarbourMasters/Shipwright/blob/807b00a058be1d9b9f19e1b787e93c6dba5803d1/soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp#L1523 to be

```cpp
if (UIWidgets::EnhancementCheckbox("Hide right-side shop item checks", "gCheckTrackerOptionHideRightShopChecks", false, "", UIWidgets::CheckboxGraphics::Cross, true)) {
    RecalculateAreaTotals();
}
```
no work for this?

Malkierian replied:
No, because it's calling the function directly, whereas the helper variables that hold whether Hide Right Side or Show GS are enabled aren't updated until the next frame, so the counts were being inverted.

Malkierian replied:
Ohhh, you know what? Maybe changing the helper variable in that if just before calling recalculate might be a better option...

briaguya0 replied:
> Ohhh, you know what? Maybe changing the helper variable in that if just before calling recalculate might be a better option...

if that works that'd be awesome!

Malkierian replied:
Yep, that means the check tracker is down to one per-frame operation, just waiting on a hook for that.

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -515,8 +542,16 @@ void CheckTrackerTransition(uint32_t sceneNum) {
+            hideShopRightChecks = hideShopRightChecks_;
+            RecalculateAreaTotals();
+        }
+        bool alwaysShowGS_ = CVarGetInteger("gCheckTrackerOptionAlwaysShowGSLocs", 0);
+        if (alwaysShowGS != alwaysShowGS_) {
+            alwaysShowGS = alwaysShowGS_;
+            RecalculateAreaTotals();
+        }
```
same question as above

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -437,8 +460,11 @@ void CheckTrackerLoadGame(int32_t fileNum) {
-            areaChecksGotten[realRcObj.rcArea]++;
+        if (IsVisibleInCheckTracker(realRcObj)) {
+            areaCheckTotals[realRcObj.rcArea]++;
+            if (rcTrackerData.status >= RCSHOW_COLLECTED || rcTrackerData.skipped) {
```
```suggestion
            if (rcTrackerData.status == RCSHOW_COLLECTED || rcTrackerData.status == RCSHOW_SAVED || rcTrackerData.skipped) {
```

briaguya0 reviewed (commented):
overall looking good, left a couple suggestions to avoid using `>`/`<` on enums (it wasn't clear what the logic was), and a couple questions about the cvar change checking per frame stuff

Malkierian committed:
Changed helper variable change and call to `RecalculateAreaTotals()` to when the options are toggled instead of checking every frame.
Removed redundant if...else.
Clarified areaChecksGotten increment/decrement functionality based on current status and incoming status change.

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -514,10 +542,18 @@ void CheckTrackerTransition(uint32_t sceneNum) {
-        hideShopRightChecks = CVarGetInteger("gCheckTrackerOptionHideRightShopChecks", 1);
-        alwaysShowGS = CVarGetInteger("gCheckTrackerOptionAlwaysShowGSLocs", 0);
-    }
+    /*if (IS_RANDO) {
+        bool hideShopRightChecks_ = CVarGetInteger("gCheckTrackerOptionHideRightShopChecks", 1);
+        if (hideShopRightChecks != hideShopRightChecks_) {
+            hideShopRightChecks = hideShopRightChecks_;
+            RecalculateAreaTotals();
+        }
+        bool alwaysShowGS_ = CVarGetInteger("gCheckTrackerOptionAlwaysShowGSLocs", 0);
+        if (alwaysShowGS != alwaysShowGS_) {
+            alwaysShowGS = alwaysShowGS_;
+            RecalculateAreaTotals();
+        }
+    }*/
```
now that this is working the other way can this part be removed instead of just commented out?

Malkierian replied:
Done.

Malkierian committed:
Removed unused code.

briaguya0 reviewed (approved):
:shipit:
